cvmfs/util/pipe.h

```diff
@@ -0,0 +1,271 @@
+   */
+  void CloseReadFd() {
+    close(fd_read_);
+    fd_read_ = -1;
+  }
+
+  /**
+   * Closes file descriptor that writes to the pipe and marks it as invalid.
+   */
+  void CloseWriteFd() {
+    close(fd_write_);
+    fd_write_ = -1;
+  }
+
+  /**
+   * Tries to write an object to the pipe
+   *
+   * @returns true if the entire object was written
+   *          false otherwise
+   */
+  template<typename T>
+  bool TryWrite(const T &data) {
+    const int num_bytes = write(fd_write_, &data, sizeof(T));
+    return (num_bytes >= 0) && (static_cast<size_t>(num_bytes) == sizeof(T));
+  }
+
+  /**
+   * Writes an object to the pipe
+   *
+   * @returns true on success
+   *          otherwise kills the program with an assert
+   *
+   */
+  template<typename T>
+  bool Write(const T &data) {
+    WritePipe(fd_write_, &data, sizeof(T));
+    return true;
+  }
+
+  /**
+   * Writes an object to the pipe
+   * If possible, it is recommended to use "bool Write(const T &data)"
+   *
+   * @returns true on success
+   *          otherwise kills the program with an assert
+   *
+   */
+  bool Write(const void *buf, size_t nbyte) {
+    WritePipe(fd_write_, buf, nbyte);
+    return true;
+  }
+
+  /**
+   * Tries to read from the pipe until it receives data or it is interrupted
+   * by a system call
+   *
+   * @returns true if sizeof(data) bytes were received
+   *          false otherwise
+   */
+  template<typename T>
+  bool TryRead(T *data) {
+    ssize_t num_bytes;
+    do {
+      num_bytes = read(fd_read_, data, sizeof(T));
+    } while ((num_bytes < 0) && (errno == EINTR));
+    return (num_bytes >= 0) && (static_cast<size_t>(num_bytes) == sizeof(T));
+  }
+
+  /**
+   * Reads an object from the pipe
+   *
+   * @returns true on success
+   *          otherwise kills the program with an assert
+   */
+  template<typename T>
+  bool Read(T *data) {
+    ReadPipe(fd_read_, data, sizeof(T));
+    return true;
+  }
+
+  /**
+   * Reads an object from the pipe
+   * If possible, it is recommend to use "bool Read(T *data)""
+   *
+   * @returns true on success
+   *          otherwise kills the program with an assert
+   */
+  bool Read(void *buf, size_t nbyte) {
+    ReadPipe(fd_read_, buf, nbyte);
+    return true;
+  }
+
+  /**
+   * Returns the file descriptor that reads from the pipe
+   */
+  int GetReadFd() {
+    return fd_read_;
+  }
+
+  /**
+   * Returns the file descriptor that writes to the pipe
+   */
+  int GetWriteFd() {
+    return fd_write_;
+  }
+
+
+ private:
+  int fd_read_;
+  int fd_write_;
+
+  /**
+   * Only used in the unit tests to test pipes using stdin/stdout as read/write.
+  */
+  Pipe(const int fd_read, const int fd_write) : fd_read_(fd_read),
+                                                fd_write_(fd_write) {}
+
+  /**
+   * Creating a pipe should always succeed.
+   */
+  void MakePipe(int pipe_fd[2]) {
+    int retval = pipe(pipe_fd);
+    if (!(retval == 0)) {
+      PANIC(kLogSyslogErr | kLogDebug,
```

Comment:
The `pipe` call can only return 0 or -1, so we don't need to print it in the error message

Reply:
so we dont want to fail if the pipe is not successfully created?

Reply:
We do, there is just no point in printing `retval`. We already know at this point that it is -1.